File: rust/crates/babylon-tick/content/scenarios/dispossession_zero_rate_conformance.py
"""The DISCRIMINATING conformance world for `dispossession/territory-transfer`.

Proves the frozen engine's exact `when`-equivalent gate — `dispossession_events.py:
75-76`: ``if foreclosure_rate <= 0.0 and eviction_rate <= 0.0 and
displacement_rate <= 0.0: continue`` — reads ONLY the three primary rates, not
`concentrated_ownership` / `absentee_landlord_share`. A territory with
nonzero structural factors but all-zero primary rates is skipped WHOLE: no
intensity computed, no state written, no event published, even though the
intensity formula would be nonzero from the structural terms alone if it
ever ran. This is not a defect the port repairs (ADR183 §5.4) — it is the
frozen system's own gate, transcribed exactly.

This scenario is also the Class C reality check
(`reports/bsl-gap-analysis-2026-08-10.md` row 10.0: "Yes (zero-rate
inputs)"): on the canonical run today nothing hydrates non-zero dispossession
rates, so this is the system's actual dormant behavior, not a hypothetical.

Run it from the repository root, single process::

    PYTHONPATH="$PWD/src" uv run python \\
        rust/crates/babylon-tick/content/scenarios/dispossession_zero_rate_conformance.py
"""

from __future__ import annotations

from typing import Any

from babylon.engine.context import TickContext
from babylon.engine.services import ServiceContainer
from babylon.engine.systems.dispossession_events import DispossessionEventSystem
from babylon.models.enums import NodeType
from babylon.topology.graph import BabylonGraph

SUBJECTS: list[tuple[str, dict[str, Any]]] = [
    (
        "dormant-county-1",
        {
            "foreclosure_rate": 0.0,
            "eviction_rate": 0.0,
            "displacement_rate": 0.0,
            # Deliberately NONZERO — proves the gate ignores these two.
            "concentrated_ownership": 0.6,
            "absentee_landlord_share": 0.4,
            "wealth": 1_000_000.0,
        },
    ),
    (
        "dormant-county-2",
        {
            "foreclosure_rate": 0.0,
            "eviction_rate": 0.0,
            "displacement_rate": 0.0,
            "concentrated_ownership": 0.6,
            "absentee_landlord_share": 0.4,
            "wealth": 500_000.0,
        },
    ),
]


def build_graph() -> BabylonGraph:
    """Build the two-territory dormant world, in scenario declaration order."""
    graph = BabylonGraph()
    for node_id, attrs in SUBJECTS:
        graph.add_node(node_id, NodeType.TERRITORY, **attrs)
    return graph


def main() -> None:
    """Run one tick of the frozen DispossessionEventSystem and print the vectors."""
    services = ServiceContainer.create()
    try:
        graph = build_graph()
        DispossessionEventSystem().step(graph, services, TickContext(tick=1))
        events = services.event_bus.get_history()

        print("post-tick state (must equal the pre-tick seed exactly — no writes):")
        for node_id, seed in SUBJECTS:
            node = graph.get_node(node_id)
            if node is None:
                raise SystemExit(f"node {node_id} vanished during the tick")
            a = node.attributes
            print(
                f"  {node_id:<18} "
                f"wealth={a['wealth']!r} "
                f"dispossession_intensity={a.get('dispossession_intensity')!r} "
                f"(seed wealth was {seed['wealth']!r})"
            )
        print()

        print("events:")
        for event in events:
            print(f"  {event.type} {event.payload!r}")
        if not events:
            print("  (none)")
    finally:
        services.database.close()


if __name__ == "__main__":
    main()

File: rust/crates/babylon-tick/tests/dispossession_zero_rate_conformance.rs
//! The DISCRIMINATING conformance suite for `dispossession/territory-transfer`
//! — proves the frozen `when`-equivalent gate is inert on the shipped
//! canonical world, exactly the way `dispossession_events.py` is today.
//!
//! # Provenance
//!
//! ```text
//! PYTHONPATH="$PWD/src" uv run python \
//!     rust/crates/babylon-tick/content/scenarios/dispossession_zero_rate_conformance.py
//! ```
//!
//! Its output on 2026-08-11, verbatim:
//!
//! ```text
//! post-tick state (must equal the pre-tick seed exactly — no writes):
//!   dormant-county-1   wealth=1000000.0 dispossession_intensity=None (seed wealth was 1000000.0)
//!   dormant-county-2   wealth=500000.0 dispossession_intensity=None (seed wealth was 500000.0)
//!
//! events:
//!   (none)
//! ```
//!
//! `dispossession_intensity=None` means the frozen engine's graph dict never
//! gained the key at all — the Rust seed uses `0` as a placeholder (BSL
//! requires every declared field to carry SOME seed value; Python's dict can
//! simply lack the key), so the assertion below checks the seed value is
//! UNCHANGED, not that it equals Python's `None`.
//!
//! `dispossession_events.py:75-76`'s `continue` reads only the three PRIMARY
//! rates (foreclosure/eviction/displacement) — `concentrated_ownership`/
//! `absentee_landlord_share` are deliberately nonzero in this scenario, and
//! the frozen engine still skips both subjects whole. That is the exact
//! fidelity claim this suite exists to prove, not an incidental detail.

use babylon_bsl::structural_verbs::CollectingSink;
use babylon_graph::memory::MemoryGraph;
use babylon_graph::substrate::{GraphSubstrate, NodeId};
use babylon_tick::{run_once, run_once_into};

const SCENARIO: &str =
    include_str!("../content/scenarios/dispossession-zero-rate-conformance.bscn");
const RULE: &str = include_str!("../content/rules/dispossession.bsl");

fn run() -> (MemoryGraph, CollectingSink) {
    let mut graph = MemoryGraph::new();
    let mut sink = CollectingSink::default();
    run_once_into(SCENARIO, RULE, &mut graph, &mut sink)
        .expect("the Dispossession pack must run, even when the guard excludes everyone");
    (graph, sink)
}

fn attribute(graph: &MemoryGraph, id: u64, field: &str) -> f64 {
    graph
        .node_attribute(NodeId(id), field)
        .unwrap_or_else(|e| panic!("node {id} field {field}: {}", e.message))
}

/// Neither subject's wealth moves: the `when` guard excludes both before any
/// binding is even read for effect purposes.
#[test]
fn no_subject_s_wealth_moves() {
    let (graph, _) = run();
    assert_eq!(attribute(&graph, 0, "territory/wealth"), 1_000_000.0);
    assert_eq!(attribute(&graph, 1, "territory/wealth"), 500_000.0);
}

/// Neither subject's intensity is written — the seed placeholder survives
/// the tick untouched, proving the gate excludes the rule's effects
/// entirely rather than computing and discarding a value.
#[test]
fn no_subject_s_intensity_is_written() {
    let (graph, _) = run();
    assert_eq!(
        attribute(&graph, 0, "territory/dispossession-intensity"),
        0.0
    );
    assert_eq!(
        attribute(&graph, 1, "territory/dispossession-intensity"),
        0.0
    );
}

/// No event fires — not `DISPOSSESSION_EVENT`, not `VALUE_TRANSFER` — for
/// either subject.
#[test]
fn no_event_fires() {
    let (_, sink) = run();
    assert!(
        sink.events.is_empty(),
        "the gate must exclude both subjects before any effect runs"
    );
}

/// The `(when …)` gate excludes both subjects: `run_tick`'s `fired` counter
/// only increments past the guard, so zero subjects fire even though two
/// are `considered`.
#[test]
fn zero_subjects_pass_the_guard() {
    let report = run_once(SCENARIO, RULE).expect("the pack must run");
    assert_eq!(report.fired, 0, "both subjects fail the primary-rate guard");
}

/// Byte-determinism, AND the strongest form of "nothing happened": the
/// pre-tick and post-tick hashes are IDENTICAL, not merely reproducible.
#[test]
fn the_tick_leaves_state_completely_unchanged() {
    let a = run_once(SCENARIO, RULE).expect("first run");
    let b = run_once(SCENARIO, RULE).expect("second run");
    assert_eq!(a.after, b.after, "two runs, one post-state");
    assert_eq!(
        a.before, a.after,
        "a tick where the guard excludes every subject must not move state at all"
    );
}
